File: DataStructures/bst.py
#!/usr/bin/python
# Author:	@BlankGodd

class Node:
	"""Binary Search Tree"""

	def __init__(self, data):
		self.data = data
		self.leftChild = None
		self.rightChild = None

class Bst:

	def __init__(self):
		self.root = None
		self.size = 0

	def insert(self, data):
		if not self.root:
			self.root = Node(data)
			self.size += 1
		else:
			self.insert_node(self.root, data)

	def insert_node(self, node, data):
		if data < node.data:
			if node.leftChild:
				self.insert_node(node.leftChild, data)
			else:
				node.leftChild = Node(data)
				self.size += 1

		else:
			if node.rightChild:
				self.insert_node(node.rightChild, data)
			else:
				node.rightChild = Node(data)
				self.size += 1

	def get_min(self):
		if not self.root:
			print("Empty Tree!")
		else:
			self.get_minimum(self.root)

	def get_minimum(self, node):
		if node.leftChild:
			self.get_minimum(node.leftChild)
		else:
			print(node.data)

	def get_max(self):
		if not self.root:
			print("Empty Tree!")
		else:
			self.get_maximum(self.root)

	def get_maximum(self, node):
		if node.rightChild:
			self.get_maximum(node.rightChild)
		else:
			print(node.data)

	def traverse(self):
		if not self.root:
			print("Empty Tree!")
		else:
			self.lst = []
			self.traverse_in_order(self.root)
			print(self.lst)

	def traverse_in_order(self, node):
		if node.leftChild:
			self.traverse_in_order(node.leftChild)
		self.lst.append(node.data)
		if node.rightChild:
			self.traverse_in_order(node.rightChild)

File: c#/bst/Bst.cs
using System;
using System.Collections.Generic;

namespace bst
{
    class Bst
    {
        public Node root;
        public List<int> tree;

        public Bst()
        {
            root  = null;
        }

        // insert
        public void insert(int data)
        {
            Node temp = new Node(data);
            if (root == null)
            {
                root = temp;
            }
            else
            {
                if (root.data < data)
                {
                    if (root.rightChild != null)
                    {
                        insertNode(root.rightChild, temp);
                    }
                    else
                    {
                        root.rightChild = temp;
                    }
                }
                else if (root.data > data)
                {
                    if (root.leftChild != null)
                    {
                        insertNode(root.leftChild, temp);
                    }
                    else
                    {
                        root.leftChild = temp;
                    }
                }
                else
                {
                    Console.WriteLine("Value already present in binary search tree!");
                }
            }
        }

        public void insertNode(Node node, Node newnode)
        {
            if (newnode.data > node.data)
            {
                if (node.rightChild == null)
                {
                    node.rightChild = newnode;
                }
                else
                {
                    insertNode(node.rightChild, newnode);
                }
            }
            else if (newnode.data < node.data)
            {
                if (node.leftChild == null)
                {
                    node.leftChild = newnode;
                }
                else
                {
                    insertNode(node.leftChild, newnode);
                }
            }
            else {
                Console.WriteLine("Value already present in binary search tree!");
            }
        }

        // get minimum
        public int getMin()
        {
            if (root == null)
            {
                Console.WriteLine("Binary search tree is empty!");
                return 0;
            }
            else
            {
                return getMinimun(root);
            }
        }

        public int getMinimun(Node node)
        {
            if (node.leftChild != null)
            {
                return getMinimun(node.leftChild);
            }
            else
            {
                return node.data;
            }
        }

        // get maximum
        public int getMax()
        {
            if (root == null)
            {
                Console.WriteLine("Binary search tree is empty!");
                return 0;
            }
            else
            {
                return getMaximum(root);
            }
        }

        public int getMaximum(Node node)
        {
            if (node.rightChild != null)
            {
                return getMaximum(node.rightChild);
            }
            else
            {
                return node.data;
            }
        }
        // traverse
        public void traverse()
        {
            tree  = new List<int>();
            if (root == null)
            {
                Console.WriteLine("Binary search tree is empty!");
            }
            else
            {
                traverseTree(root);
            }
            foreach (int item in tree)
            {
                Console.WriteLine(item);
            }
        }

        public void traverseTree(Node node)
        {
            if (node.leftChild != null)
            {
                traverseTree(node.leftChild);
            }

            tree.Add(node.data);

            if (node.rightChild != null)
            {
                traverseTree(node.rightChild);
            }
        }
        // search
    }
}